[Offload] Use new error code handling mechanism and lower-case messages

RossBrunton opened the pull request:
[Offload] Use new error code handling mechanism

This removes the old ErrorCode-less error method and requires
every user to provide a concrete error code. All calls have been
updated.

In addition, for consistency with error messages elsewhere in LLVM, all messages have been made to start lower case.

RossBrunton committed:
[Offload] Use new error code handling mechanism

This removes the old ErrorCode-less error method and requires
every user to provide a concrete error code. All calls have been
updated.

jhuber6 commented on offload/plugins-nextgen/common/src/GlobalHandler.cpp:
```diff
@@ -32,7 +32,16 @@ GenericGlobalHandlerTy::getELFObjectFile(DeviceImageTy &Image) {
-  return ELFObjectFileBase::createELFObjectFile(Image.getMemoryBuffer());
+  auto Expected =
+      ELFObjectFileBase::createELFObjectFile(Image.getMemoryBuffer());
+  if (!Expected) {
+    return handleErrors(Expected.takeError(), [](llvm::StringError &Err) {
+      return llvm::createStringError(ErrorCode::INVALID_BINARY,
+                                     "Error parsing binary: " +
+                                         Err.getMessage());
+    });
+  }
+  return Expected;
```
Can we make this a common utility? Seems like it may show up in other places.

RossBrunton replied:
I've added a new overload of `Plugin::error` which accepts an `llvm::Error` and a context. If it is given a non-string error, it just has `Context` as its error message. If it is a string error, its error message is in the form `${Context}: ${Message}`.

jhuber6 reviewed (approved):
This is a really good start, thanks.

arsenm commented on offload/include/Shared/OffloadErrcodes.inc:
```diff
@@ -0,0 +1,51 @@
+// To add new error codes, add them to offload/liboffload/API/Common.td and run
+// the GenerateOffload target.
+
+OFFLOAD_ERRC(SUCCESS, "Success", 0)
```
Error messages are supposed to start with a lowercase letter

RossBrunton replied:
I don't disagree with that, but there seems to be precedent in offload for error messages to start with a capital letter.

Since I'm touching error messages anyway, is it worth me just lower casing them now?

jhuber6 replied:
Sure

RossBrunton replied:
Done.

While doing the search, I also noticed a few places where I've missed error reporting, so those have been updated as well. Which meant moving/adding helper functions to offloadError.h .

arsenm commented on offload/include/Shared/OffloadErrcodes.inc:
```diff
@@ -0,0 +1,51 @@
+//===- Auto-generated file, part of the LLVM/Offload project --------------===//
```
Why is it being committed if it's generated?

RossBrunton replied:
This file is from a change that this PR depends on, #138258 .

But basically the reasoning is that a lot of liboffload consists of generated code that is committed into the repo. Since it uses the error codes, it makes sense for them to also be committed otherwise they could get out of sync if only one half gets regen-ed.

shiltian replied:
I think @arsenm 's point is, why not just generate it on the fly during building, like other tablegen files in LLVM.

callumfare replied:
So with the original patch that landed liboffload (#108413) we reached the decision to check in the auto-generated files. The reasons we had for doing this (in descending order of importance) were:
- The entire new Offload API header is generated. If we didn't check it in, there would be a whole standalone library intended to be used by runtimes and other users that had no headers in the repo. This seemed like it would be a major source of a confusion.
- It made it easier for reviewers to follow what the offload-tblgen tooling is doing, both in that PR and in follow up changes
- Realistically we need clang-format to ensure the autogenerated code is nicely formatted and there were some concerns that it wouldn't always be available in the build. This is also why the `OffloadGenerate` target isn't a hard dependency of the build.

There was a lot of discussion on the pros/cons of this. Obviously this isn't how tablegen is used elsewhere and it's not necessarily intuitive to other LLVM devs. It also invites situations where the `.td` files are updated but somebody forgets to check in the newest generated files (this has already happened once).

I think the latter two bullet points aren't really a big concern at this point, but the first one still is IMO.

arsenm replied:
Point 1 is not really a point of confusion, LLVM has no shortage of generated public headers. I also don't think it's important to have nicely formatted generated code, but you could always add a clang-format to the build steps

RossBrunton committed:
Lowercase error messages and port over more calls

Error messages should be lower case, so all of them have been changed.
In addition, a few more string error creations have been ported to
use offload errors, which also meant moving some reporting things to
OffloadError.h.

RossBrunton committed:
Formatting

RossBrunton committed:
Handle not found for cuda
